Receive payloads on client stream

deepthidevaki opened the pull request:
## Description

This PR implements the first part of #11712  - adding support for receiving jobs in the transport.

`ClientStreamService` adds a subscriber to handle push requests from servers. There is only one push request topic for all streams. `ClientStreamService` forwards payload to the respective client stream based on the provided streamId in the push request. 

We assume the `ClientStreamConsumer` does the error handling itself. For example, if the grpc stream is already closed, the consumer should take care of making the job re-activatable. Any other unexpected exceptions when invoking the consumer will be forwarded back to the server. 

A integration to test interoperability of client side and server side stream management is also added.

## Related issues

related #11712 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

npepinpe commented on transport/src/test/java/io/camunda/zeebe/transport/stream/impl/TestCommunicationService.java:
```diff
@@ -0,0 +1,172 @@
+import java.util.function.Function;
+
+/** A mock implementation of ClusterCommunicationService */
+final class TestCommunicationService implements ClusterCommunicationService {
```
:rocket:

deepthidevaki replied:
This is now added in transport. But if it is useful in other tests, we could also move it to reuse it..

deepthidevaki commented on transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamManager.java:
```diff
@@ -63,4 +62,19 @@ void remove(final UUID streamId) {
+      clientStream.getClientStreamConsumer().push(payload);
+      return true;
+    } else {
+      // client stream is closed
+      // TODO: remove stream from broker or return an error so that broker can close the stream
+      return false;
```
:question: How to handle this case? 
- Return an error, Broker on receiving this error can remove the stream its registry.
- Client sends a explicit remove stream request

deepthidevaki commented on transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamService.java:
```diff
@@ -29,13 +32,30 @@ public ClientStreamService(final ClusterCommunicationService communicationServic
+  protected void onActorStarted() {
+    // TODO: define a response type, so that broker can detect success or failure and take required
+    // action
+    communicationService.subscribe(
+        StreamTopics.PUSH.topic(),
+        MessageUtil::parsePushRequest,
+        clientStreamManager::onPayloadReceived,
+        actor::run);
```
TODO: Here we may have to define a response type and send a response. Otherwise, it seems the handler is just a consumer and do not send exception back to the sender in case something fails.

npepinpe commented on transport/src/test/java/io/camunda/zeebe/transport/stream/impl/TestCommunicationService.java:
```diff
@@ -0,0 +1,180 @@
+      final Function<M, byte[]> encoder,
+      final MemberId memberId,
+      final boolean reliable) {
+    // Do nothing
```
:wrench: If this is not implemented because we're not testing it yet, then let's add `throw new UnsupportedOperationException`. Otherwise I don't understand why it's implemented to do nothing :thinking:

deepthidevaki replied:
We are using unicast for sending remove requests. But since we are not waiting for ack, there is no need to implement the actual request handling. I will update the comment.

npepinpe replied:
If I wanted to run an integration test using this, and I wanted to test that the receiver did receive the request anyway, wouldn't I need to implement something here?

deepthidevaki replied:
Then, it's better to implement it. We cannot throw UnsupportedException because the method is used.

npepinpe reviewed (approved):
:rocket: 

:thought_balloon: Based on your comment:

> ClientStreamService forwards payload to the respective client stream based on the provided streamId in the push request.

We should at least add one or more metrics around that to detect congestion, as it could be a choke point.

:question: Regarding this comment:

> We assume the ClientStreamConsumer does the error handling itself. For example, if the grpc stream is already closed, the consumer should take care of making the job re-activatable

I expect this is a race condition? Ideally, if the gRPC stream is closed, then the client is also closed, and we wouldn't push to it. Is this what you meant?

This makes me think we need to add a stateless error handler, most likely, for cases where there are no streams registered on our side, but we still want to yield the job back.

deepthidevaki commented on transport/src/main/java/io/camunda/zeebe/transport/stream/api/ClientStreamConsumer.java:
```diff
@@ -7,21 +7,21 @@
  * @param <P> the payload type that can be pushed to the stream
  */
-public interface ClientStreamConsumer<P extends BufferReader> {
+public interface ClientStreamConsumer {
```
Just noticed, reference to type parameter has to be removed from its javadocs.

deepthidevaki commented on transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamManager.java:
```diff
@@ -57,10 +57,30 @@ UUID add(
+          // Stream does not exist. We expect to have already sent remove request to all servers.
+          // But just in case that request is lost, we send remove request again. To keep it simple,
+          // we do not retry. Otherwise, it is possible that we send it multiple times unnecessary.
+          requestManager.removeStreamUnreliable(streamId, servers);
```
The exception will be propagated back to the broker. The broker will pass the error back to the engine, which can make the job re-activatable. As we discussed, there are some advantages to handle this error in the gateway itself. I would suggest to address it in a second pass.

deepthidevaki commented on transport/src/main/java/io/camunda/zeebe/transport/stream/impl/ClientStreamService.java:
```diff
@@ -29,13 +33,41 @@ public ClientStreamService(final ClusterCommunicationService communicationServic
+
+  @Override
+  protected void onActorStarted() {
+    // TODO: Define an PushResponse to inform server if push was successful or not. Currently, an
```
The exception is propagated back to the broker. But if we want to provide actionable error codes, then it would make sense to define a proper error response. I think it is also cleaner to do it, compared to sending the whole exception stack trace. But it is not critical. The current state achieves what we want to do - error is received by the broker, will be propagated back to the engine.

deepthidevaki commented on transport/src/test/java/io/camunda/zeebe/transport/stream/impl/StreamIntegrationTest.java:
Using ControlledActorScheduler is more cumbersome. We have to insert `workUntilDone` at several places to force it to execute the tasks. I find it unnecessary complexity for this test. If we add more test cases where we need fine grained concurrency control, we can move the test to use ControlledActorScheduler then.

npepinpe replied:
It's a shame. Wish there was a way to test two actors in isolation without having to deal with asynchronicity and potential flakiness :(

deepthidevaki commented:
I have addressed most of your comments:
- Let's add metrics in a followup PR.
- As discussed, let's follow up if we want to provide a stateless error handler to handle the case were no stream is found. Currently, the error is propagated back to the broker, which will be then handled by engine. It can then mark the job as re-activatable.

deepthidevaki commented:
bors merge
